README.rst
==========
MonthDelta
==========

A single module, |mod|_, is provided by the |MonthDelta|_ distribution. The
|mod|_ module includes a class, |MD|_, and a function, `monthmod()`_. |MD|_
enables easy month-related calculations with the standard Python date_ and
datetime_ classes from the |DT|_ module. `monthmod()`_ enables round-trip
calculations among |MD|_, date_, datetime_, and timedelta_. Starting with
`version 2.3`_ (when the |DT|_ module was added), |MonthDelta|_ supports all
subsequent versions of Python_ from one codebase.

This library was originally developed as an enhancement to the C implementation
of Python_, to be included directly in the |DT|_ module. That was not seen as a
valuable addition to the standard library, so this "back-ported" Python_ module
is the way to use this functionality.

Documentation_ is available, and the source_ is hosted at Bitbucket_. `Jess
Austin`_ is the author of |MonthDelta|_. |MonthDelta|_ is distributed under
the `MIT license`_.

.. |MonthDelta| replace:: **MonthDelta**
.. |mod| replace:: **monthdelta**
.. _mod: http://packages.python.org/MonthDelta/#module-monthdelta
.. |MD| replace:: MonthDelta
.. _MD: http://packages.python.org/MonthDelta#monthdelta.MonthDelta
.. _`monthmod()`: http://packages.python.org/MonthDelta#monthdelta.monthmod
.. |DT| replace:: **datetime**
.. _DT: http://docs3.python.org/library/datetime.html
.. _date: http://docs3.python.org/library/datetime.html#datetime.date
.. _datetime: http://docs3.python.org/library/datetime.html#datetime.datetime
.. _timedelta: http://docs3.python.org/library/datetime.html#datetime.timedelta
.. _Python: http://python.org
.. _Documentation: http://packages.python.org/MonthDelta
.. _source: https://bitbucket.org/jessaustin/monthdelta
.. _Bitbucket: https://bitbucket.org
.. _`Jess Austin`: mailto:jess.austin@gmail.com
.. _`MIT License`: http://www.opensource.org/licenses/mit-license.php
.. _`version 2.3`: http://www.python.org/download/releases/2.3.7


CHANGES.txt
v1.0a, June 2012 -- cleaned up PyPI stuff; pep8 camelcase for monthdelta class;
                    no more max/min.


test_monthdelta.py
"""Test date/time type.

See http://www.zope.org/Members/fdrake/DateTimeWiki/TestCases
"""

import os
import pickle
import unittest
import types

from operator import lt, le, gt, ge, eq, ne

try:
    reversed
    sorted
except NameError:
    def reversed(seq):
        return seq[::-1]
    def sorted(it):
        it = list(it)
        it.sort()
        return it
try:
    from test import support
except ImportError:
    from test import test_support as support
try:
    from itertools import combinations, permutations
except ImportError:
    def permutations(iterable, r=None):
        pool = tuple(iterable)
        n = len(pool)
        if r is None:
            r = n
        indices = range(n)
        cycles = range(n, n-r, -1)
        yield tuple([pool[i] for i in indices[:r]])
        while n:
            for i in reversed(range(r)):
                cycles[i] -= 1
                if cycles[i] == 0:
                    indices[i:] = indices[i+1:] + indices[i:i+1]
                    cycles[i] = n - i
                else:
                    j = cycles[i]
                    indices[i], indices[-j] = indices[-j], indices[i]
                    yield tuple([pool[i] for i in indices[:r]])
                    break
            else:
                return
    def combinations(iterable, r):
        pool = tuple(iterable)
        n = len(pool)
        for indices in permutations(range(n), r):
            if sorted(indices) == list(indices):
                yield tuple([pool[i] for i in indices])

from datetime import MINYEAR, MAXYEAR
from datetime import timedelta
from datetime import tzinfo
from datetime import time
from datetime import date, datetime
from monthdelta import monthdelta, monthmod

pickle_choices = [(pickle, pickle, proto) for proto in range(3)]
assert len(pickle_choices) == 3

# An arbitrary collection of objects of non-datetime types, for testing
# mixed-type comparisons.
OTHERSTUFF = (10, 10, 34.5, "abc", {}, [], ())

class TestMonthDelta(unittest.TestCase):
    expectations = (
        (date(2006,12,31), monthdelta(6),   date(2007,6,30),date(2006,6,30)),
        (date(2007,1,1),   monthdelta(6),   date(2007,7,1), date(2006,7,1)),
        (date(2007,1,2),   monthdelta(6),   date(2007,7,2), date(2006,7,2)),
        (date(2006,12,31), monthdelta(12),date(2007,12,31),date(2005,12,31)),
        (date(2007,1,1),   monthdelta(12),  date(2008,1,1),date(2006,1,1)),
        (date(2007,1,2),   monthdelta(12),  date(2008,1,2),date(2006,1,2)),
        (date(2006,12,31), monthdelta(60),date(2011,12,31),date(2001,12,31)),
        (date(2007,1,1),   monthdelta(60),  date(2012,1,1),date(2002,1,1)),
        (date(2007,1,2),   monthdelta(60),  date(2012,1,2),date(2002,1,2)),
        (date(2006,12,31),monthdelta(600),date(2056,12,31),date(1956,12,31)),
        (date(2007,1,1),   monthdelta(600), date(2057,1,1),date(1957,1,1)),
        (date(2007,1,2),   monthdelta(600), date(2057,1,2),date(1957,1,2)),
        (date(2007,2,27), monthdelta(1), date(2007, 3, 27),date(2007,1, 27)),
        (date(2007,2,28), monthdelta(1), date(2007, 3, 28),date(2007,1, 28)),
        (date(2007,3,1),  monthdelta(1), date(2007, 4, 1), date(2007, 2, 1)),
        (date(2007,3,30), monthdelta(1), date(2007, 4, 30),date(2007,2, 28)),
        (date(2007,3,31), monthdelta(1), date(2007, 4, 30),date(2007,2, 28)),
        (date(2007,4,1),  monthdelta(1), date(2007, 5, 1), date(2007, 3, 1)),
        (date(2008,2,27), monthdelta(1), date(2008, 3, 27),date(2008,1, 27)),
        (date(2008,2,28), monthdelta(1), date(2008, 3, 28),date(2008,1, 28)),
        (date(2008,2,29), monthdelta(1), date(2008, 3, 29),date(2008,1, 29)),
        (date(2008,3,1),  monthdelta(1), date(2008, 4, 1), date(2008, 2, 1)),
        (date(2008,3,30), monthdelta(1), date(2008, 4, 30),date(2008,2, 29)),
        (date(2008,3,31), monthdelta(1), date(2008, 4, 30),date(2008,2, 29)),
        (date(2008,4,1),  monthdelta(1), date(2008, 5, 1), date(2008, 3, 1)),
        (date(2100,2,27), monthdelta(1), date(2100, 3, 27),date(2100,1, 27)),
        (date(2100,2,28), monthdelta(1), date(2100, 3, 28),date(2100,1, 28)),
        (date(2100,3,1),  monthdelta(1), date(2100, 4, 1), date(2100, 2, 1)),
        (date(2100,3,30), monthdelta(1), date(2100, 4, 30),date(2100,2, 28)),
        (date(2100,3,31), monthdelta(1), date(2100, 4, 30),date(2100,2, 28)),
        (date(2100,4,1),  monthdelta(1), date(2100, 5, 1), date(2100, 3, 1)),
        (date(2000,2,27), monthdelta(1), date(2000, 3, 27),date(2000,1, 27)),
        (date(2000,2,28), monthdelta(1), date(2000, 3, 28),date(2000,1, 28)),
        (date(2000,2,29), monthdelta(1), date(2000, 3, 29),date(2000,1, 29)),
        (date(2000,3,1),  monthdelta(1), date(2000, 4, 1), date(2000, 2, 1)),
        (date(2000,3,30), monthdelta(1), date(2000, 4, 30),date(2000,2, 29)),
        (date(2000,3,31), monthdelta(1), date(2000, 4, 30),date(2000,2, 29)),
        (date(2000,4,1),  monthdelta(1), date(2000, 5, 1), date(2000, 3, 1)))

    def test_calc(self):
        for dt, md, sub, prev in self.expectations:
            self.assertEqual(dt + md, sub)
            self.assertEqual(dt - md, prev)
    def test_math(self):
        for x, y in permutations(range(26),2):
            self.assertEqual(monthdelta(x) + monthdelta(y), monthdelta(x +y))
            self.assertEqual(monthdelta(x) - monthdelta(y), monthdelta(x -y))
            self.assertEqual(monthdelta(x) * y, monthdelta(x * y))
        for x, y in combinations(range(26),2):
            self.assertEqual(monthdelta(x) // y, monthdelta(x // y))
            self.assertEqual(monthdelta(x) // monthdelta(y), x // y)
    def test_comp(self):
        for x, y in combinations(range(26),2):
            self.assert_(monthdelta(x) < monthdelta(y))
            self.assert_(monthdelta(x) <= monthdelta(y))
            self.assert_(monthdelta(x) != monthdelta(y))
            self.assert_(monthdelta(y) > monthdelta(x))
            self.assert_(monthdelta(y) >= monthdelta(x))
        for x in range(26):
            self.assert_(monthdelta(x) <= monthdelta(x))
            self.assert_(monthdelta(x) == monthdelta(x))
            self.assert_(monthdelta(x) >= monthdelta(x))
    def test_bool(self):
        self.assert_(monthdelta())
        self.assert_(monthdelta(-1))
        self.assert_(not monthdelta(0))
    def test_class(self):
        self.assertEqual(monthdelta.min, monthdelta(-99999999))
        self.assertEqual(monthdelta.max, monthdelta(99999999))
    def test_subclass(self):
        class M(monthdelta):
            def from_md(md):
                return M(md.months)
            from_md = staticmethod(from_md)
            def as_years(self):
                return round(self.months / 12)

        m1 = M()
        self.assert_(type(m1) is M or type(m1) is types.InstanceType)
        self.assertEqual(m1.as_years(), 0)
        m2 = M(-24)
        self.assert_(type(m2) is M or type(m2) is types.InstanceType)
        self.assertEqual(m2.as_years(), -2)
        m3 = m1 + m2
        self.assert_(type(m3) is monthdelta or type(m3) is types.InstanceType)
        m4 = M.from_md(m3)
        self.assert_(type(m4) is M or type(m4) is types.InstanceType)
        self.assertEqual(m3.months, m4.months)
        self.assertEqual(str(m3), str(m4))
        self.assertEqual(m4.as_years(), -2)
    def test_str(self):
        self.assertEqual(str(monthdelta()), '1 month')
        self.assertEqual(str(monthdelta(-1)), '-1 month')
        self.assertEqual(str(monthdelta(3)), '3 months')
        self.assertEqual(str(monthdelta(-17)), '-17 months')
    def test_pickling(self):
        orig = monthdelta(42)
        green = pickle.dumps(orig)
        derived = pickle.loads(green)
        self.assertEqual(orig, derived)
    def test_disallowed(self):
        a = monthdelta(42)
        for i in 1, 1.0:
            self.assertRaises(TypeError, lambda: a+i)
            self.assertRaises(TypeError, lambda: a-i)
            self.assertRaises(TypeError, lambda: i+a)
            self.assertRaises(TypeError, lambda: i-a)
            self.assertRaises(TypeError, lambda: a/i)
            self.assertRaises(TypeError, lambda: i/a)
            self.assertRaises(TypeError, lambda: a/a)
            self.assertRaises(ZeroDivisionError, lambda: a // 0)
        def inplace_fail():
            b = monthdelta(12)
            b //= monthdelta(3)
        self.assertRaises(TypeError, inplace_fail)
        x = 2.3
        self.assertRaises(TypeError, lambda: a*x)
        self.assertRaises(TypeError, lambda: x*a)
        self.assertRaises(TypeError, lambda: a // x)
        self.assertRaises(TypeError, lambda: x // a)
        self.assertRaises(OverflowError, monthdelta, -100000000)
        self.assertRaises(OverflowError, monthdelta, 100000000)

class TestMonthMod(unittest.TestCase):
    md_zero, td_zero = monthdelta(0), timedelta(0)
    expectations = (
        (date(2007,1,1),  date(2007,1,1),  md_zero, td_zero),
        (date(2007,2,28), date(2007,2,28), md_zero, td_zero),
        (date(2007,3,1),  date(2007,3,1),  md_zero, td_zero),
        (date(2008,2,28), date(2008,2,28), md_zero, td_zero),
        (date(2008,2,29), date(2008,2,29), md_zero, td_zero),
        (date(2008,3,1),  date(2008,3,1),  md_zero, td_zero),
        (date(2007,1,1),  date(2007,2,27), monthdelta(1),  timedelta(26)),
        (date(2007,1,1),  date(2007,2,28), monthdelta(1),  timedelta(27)),
        (date(2007,1,1),  date(2007,3,1),  monthdelta(2),  timedelta(0)),
        (date(2007,1,1),  date(2007,3,30), monthdelta(2),  timedelta(29)),
        (date(2007,1,1),  date(2007,3,31), monthdelta(2),  timedelta(30)),
        (date(2007,1,1),  date(2007,4,1),  monthdelta(3),  timedelta(0)),
        (date(2008,1,1),  date(2008,2,27), monthdelta(1),  timedelta(26)),
        (date(2008,1,1),  date(2008,2,28), monthdelta(1),  timedelta(27)),
        (date(2008,1,1),  date(2008,2,29), monthdelta(1),  timedelta(28)),
        (date(2008,1,1),  date(2008,3,1),  monthdelta(2),  timedelta(0)),
        (date(2008,1,1),  date(2008,3,30), monthdelta(2),  timedelta(29)),
        (date(2008,1,1),  date(2008,3,31), monthdelta(2),  timedelta(30)),
        (date(2008,1,1),  date(2008,4,1),  monthdelta(3),  timedelta(0)),
        (date(2006,1,1),  date(2007,2,27), monthdelta(13), timedelta(26)),
        (date(2006,1,1),  date(2007,2,28), monthdelta(13), timedelta(27)),
        (date(2006,1,1),  date(2007,3,1),  monthdelta(14), timedelta(0)),
        (date(2006,1,1),  date(2007,3,30), monthdelta(14), timedelta(29)),
        (date(2006,1,1),  date(2007,3,31), monthdelta(14), timedelta(30)),
        (date(2006,1,1),  date(2007,4,1),  monthdelta(15), timedelta(0)),
        (date(2006,1,1),  date(2008,2,27), monthdelta(25), timedelta(26)),
        (date(2006,1,1),  date(2008,2,28), monthdelta(25), timedelta(27)),
        (date(2006,1,1),  date(2008,2,29), monthdelta(25), timedelta(28)),
        (date(2006,1,1),  date(2008,3,1),  monthdelta(26), timedelta(0)),
        (date(2006,1,1),  date(2008,3,30), monthdelta(26), timedelta(29)),
        (date(2006,1,1),  date(2008,3,31), monthdelta(26), timedelta(30)),
        (date(2006,1,1),  date(2008,4,1),  monthdelta(27), timedelta(0)),
        (date(2007,2,27), date(2007,1,1),  monthdelta(-2), timedelta(5)),
        (date(2007,2,28), date(2007,1,1),  monthdelta(-2), timedelta(4)),
        (date(2007,3,1),  date(2007,1,1),  monthdelta(-2), timedelta(0)),
        (date(2007,3,30), date(2007,1,1),  monthdelta(-3), timedelta(2)),
        (date(2007,3,31), date(2007,1,1),  monthdelta(-3), timedelta(1)),
        (date(2007,4,1),  date(2007,1,1),  monthdelta(-3), timedelta(0)),
        (date(2008,2,27), date(2008,1,1),  monthdelta(-2), timedelta(5)),
        (date(2008,2,28), date(2008,1,1),  monthdelta(-2), timedelta(4)),
        (date(2008,2,29), date(2008,1,1),  monthdelta(-2), timedelta(3)),
        (date(2008,3,1),  date(2008,1,1),  monthdelta(-2), timedelta(0)),
        (date(2008,3,30), date(2008,1,1),  monthdelta(-3), timedelta(2)),
        (date(2008,3,31), date(2008,1,1),  monthdelta(-3), timedelta(1)),
        (date(2008,4,1),  date(2008,1,1),  monthdelta(-3), timedelta(0)),
        (date(2007,2,27), date(2006,1,1),  monthdelta(-14), timedelta(5)),
        (date(2007,2,28), date(2006,1,1),  monthdelta(-14), timedelta(4)),
        (date(2007,3,1),  date(2006,1,1),  monthdelta(-14), timedelta(0)),
        (date(2007,3,30), date(2006,1,1),  monthdelta(-15), timedelta(2)),
        (date(2007,3,31), date(2006,1,1),  monthdelta(-15), timedelta(1)),
        (date(2007,4,1),  date(2006,1,1),  monthdelta(-15), timedelta(0)),
        (date(2008,2,27), date(2006,1,1),  monthdelta(-26), timedelta(5)),
        (date(2008,2,28), date(2006,1,1),  monthdelta(-26), timedelta(4)),
        (date(2008,2,29), date(2006,1,1),  monthdelta(-26), timedelta(3)),
        (date(2008,3,1),  date(2006,1,1),  monthdelta(-26), timedelta(0)),
        (date(2008,3,30), date(2006,1,1),  monthdelta(-27), timedelta(2)),
        (date(2008,3,31), date(2006,1,1),  monthdelta(-27), timedelta(1)),
        (date(2008,4,1),  date(2006,1,1),  monthdelta(-27), timedelta(0)),
        (date.min, date.max-timedelta(365), monthdelta(119975), timedelta(30)))

    def test_calc(self):
        for start, end, md, td in self.expectations:
            self.assertEqual(monthmod(start, end), (md, td))
            self.assert_((start > end and md < self.md_zero) or
                         (start <= end and md >= self.md_zero))
            self.assert_(td >= self.td_zero)
            self.assert_(td < end.replace(end.year+end.month//12,
                                          end.month%12+1, 1) -
                              end.replace(day=1))
    def test_invariant(self):
        for start, end, md, td in self.expectations:
            self.assertEqual(sum(monthmod(start, start + td), start),
                             start + td)
            self.assertEqual(sum(monthmod(end, end + td), end),
                             end + td)

    def test_error_handling(self):
        self.assertRaises(TypeError, monthmod, date.min)
        self.assertRaises(TypeError, monthmod, 123, 'abc')
        self.assertRaises(TypeError, monthmod, end=date.max)
        self.assertRaises(TypeError, monthmod, date.min, datetime.max)
        self.assertRaises(TypeError, monthmod, datetime.min, date.max)
        # perhaps it would be better not to overflow for this, but we rely on
        # the addition defined by the type of the arguments
        self.assertRaises(OverflowError, monthmod, date.min+timedelta(1),
                                                   date.min)

def test_main():
    support.run_unittest(TestMonthDelta, TestMonthMod)

if __name__ == "__main__":
    test_main()


monthdelta.py
"""monthdelta

Date calculation with months: MonthDelta class and monthmod() function.
"""

__all__ = ['MonthDelta', 'monthmod']

from datetime import date, timedelta

class MonthDelta:
    """Number of months offset from a date or datetime.

    MonthDeltas allow date calculation without regard to the different lengths
    of different months. A MonthDelta value added to a date produces another
    date that has the same day-of-the-month, regardless of the lengths of the
    intervening months. If the resulting date is in too short a month, the
    last day in that month will result:

        date(2008,1,30) + MonthDelta(1) -> date(2008,2,29)

    MonthDeltas may be added, subtracted, multiplied, and floor-divided
    similarly to timedeltas. They may not be added to timedeltas directly, as
    both classes are intended to be used directly with dates and datetimes.
    Only ints may be passed to the constructor. MonthDeltas are immutable.

    NOTE: in calculations involving the 29th, 30th, and 31st days of the
    month, MonthDeltas are not necessarily invertible [i.e., the result above
    would not imply that date(2008,2,29) - MonthDelta(1) -> date(2008,1,30)].
    """
    __slots__ = ('__months',)

    def __init__(self, months=1):
        if not isinstance(months, int):
            raise TypeError('months must be an integer')
        self.__months = months
    def months(self):
        return self.__months
    months = property(months)
    def __repr__(self):
        try:
            return 'MonthDelta({0})'.format(self.__months)
        except AttributeError:
            return 'MonthDelta(' + str(self.__months) + ')'
    def __str__(self):
        return str(self.__months) + ' month' + ((abs(self.__months) != 1
                                                 and 's') or '')
    def __hash__(self):
        return hash(self.__months)
    def __eq__(self, other):
        if isinstance(other, MonthDelta):
            return (self.__months == other.months)
        return False
    def __ne__(self, other):
        if isinstance(other, MonthDelta):
            return (self.__months != other.months)
        return True
    def __lt__(self, other):
        if isinstance(other, MonthDelta):
            return (self.__months < other.months)
        return NotImplemented
    def __le__(self, other):
        if isinstance(other, MonthDelta):
            return (self.__months <= other.months)
        return NotImplemented
    def __gt__(self, other):
        if isinstance(other, MonthDelta):
            return (self.__months > other.months)
        return NotImplemented
    def __ge__(self, other):
        if isinstance(other, MonthDelta):
            return (self.__months >= other.months)
        return NotImplemented
    def __add__(self, other):
        if isinstance(other, MonthDelta):
            return MonthDelta(self.__months + other.months)
        if isinstance(other, date):
            day = other.day
            # subract one because months are not zero-based
            month = other.month + self.__months - 1
            year = other.year + month // 12
            # now add it back
            month = month % 12 + 1
            if month == 2:
                if day >= 29 and not year%4 and (year%100 or not year%400):
                    day = 29
                elif day > 28:
                    day = 28
            elif month in (4,6,9,11) and day > 30:
                day = 30
            try:
                return other.replace(year, month, day)
            except ValueError:
                raise OverflowError('date value out of range')
        return NotImplemented
    def __sub__(self, other):
        if isinstance(other, MonthDelta):
            return MonthDelta(self.__months - other.months)
        return NotImplemented
    def __mul__(self, other):
        if isinstance(other, int):
            return MonthDelta(self.__months * other)
        return NotImplemented
    def __floordiv__(self, other):
        # MonthDelta // MonthDelta -> int
        if isinstance(other, MonthDelta):
            return self.__months // other.months
        if isinstance(other, int):
            return MonthDelta(self.__months // other)
        return NotImplemented
    def __radd__(self, other):
        return self + other
    def __rsub__(self, other):
        return -self + other
    def __rmul__(self, other):
        return self * other
    def __ifloordiv__(self, other):
        # in-place division by a MonthDelta (which will change the variable's
        # type) is almost certainly a bug -- raising this error is the reason
        # we don't just fall back on __floordiv__
        if isinstance(other, MonthDelta):
            raise TypeError('in-place division of a MonthDelta requires an '
                            'integer divisor')
        if isinstance(other, int):
            return MonthDelta(self.__months // other)
        return NotImplemented
    def __neg__(self):
        return MonthDelta(-self.__months)
    def __pos__(self):
        return MonthDelta(+self.__months)
    def __abs__(self):
        return MonthDelta(abs(self.__months))
    def __bool__(self):
        return bool(self.__months)
    __nonzero__ = __bool__
    

def monthmod(start, end):
    """Months between dates, plus leftover time.

    Distribute the interim between start and end dates into MonthDelta and
    timedelta portions. If and only if start is after end, returned MonthDelta
    will be negative. Returned timedelta is always non-negative, and is always
    smaller than the month in which the end date occurs.

    Invariant: dt + monthmod(dt, dt+td)[0] + monthmod(dt, dt+td)[1] = dt + td
    """
    if not (isinstance(start, date) and isinstance(end, date)):
        raise TypeError('start and end must be dates')
    md = MonthDelta(12*(end.year - start.year) + end.month - start.month -
                    int(start.day > end.day))
    # will overflow (underflow?) for end near date.min
    return md, end - (start + md)
